### PR TITLE
Feature/18/add prometheus support

### PR DESCRIPTION
**Pull request by svettwer**

### CODE REVIEW

File: packages/sakuli-forwarder-prometheus/src/prometheus-forwarder.class.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            addStepCriticalThresholdGauge(testStepIndex, testStepContext);
+            addStepDurationGauge(testCaseIndex, testCaseContext, testStepIndex, testStepContext);
+        }
+    }
+
+    private isNotLegacyTestStep(testStepContext: TestContextEntity,
+                             testStepIndex: number,
+                             testCaseContext: TestContextEntity) {
+        return testStepContext.id || testStepIndex !== testCaseContext.getChildren().length-1;
+    }
```
Comment (s1hofmann) [inline packages/sakuli-forwarder-prometheus/src/prometheus-forwarder.class.ts]: I'm a bit confused by the name here. What exactly is a legacy teststep? Why do we check for this property?

Also, I'd expect return type `boolean` for `is...` kind of methods / functions, not `string | boolean`.

File: packages/sakuli-forwarder-prometheus/src/gauge.utils.ts [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+function createStepIdentifier(testStepIndex: number, testStepContext: TestContextEntity){
+    return `${addPaddingZeroes(testStepIndex)}_${getEntityId(testStepContext)}`
+}
+
+function createActionIdentifier(testActionIndex: number, testActionContext: TestContextEntity) {
+    return `${addPaddingZeroes(testActionIndex)}_${getEntityId(testActionContext)}`;
+}
+
+function  addPaddingZeroes(number: number){
+    return number.toString().padStart(3, '0');
```
Comment (s1hofmann) [inline packages/sakuli-forwarder-prometheus/src/gauge.utils.ts:201]: `String.prototype.padStart` is part of `ES2017`. With current tsconfig settings, TypeScript complains about not knowing `padStart`, but I guess we can upgrade our tsconfig to ES2017. See https://node.green/#ES2017-features-String-padding-String-prototype-padStart

File: packages/sakuli-forwarder-prometheus/src/prometheus-forwarder.class.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        ifError(testCaseContext, () => addCaseError(testSuiteContext, testCaseIndex, testCaseContext));
+        addCaseDurationGauge(testSuiteContext, testCaseIndex, testCaseContext);
+        addCaseWarningThresholdGauge(testCaseIndex, testCaseContext);
+        addCaseCriticalThresholdGauge(testCaseIndex, testCaseContext);
+    }
+
+    private addStepGauges(testStepContext: TestContextEntity,
+                          testCaseIndex: number,
+                          testCaseContext: TestContextEntity,
+                          testStepIndex: number) {
```
Comment (s1hofmann) [inline packages/sakuli-forwarder-prometheus/src/prometheus-forwarder.class.ts]: Ordering of parameters is a bit off here, step and case are interleaved.

File: packages/sakuli-forwarder-prometheus/src/gauge.utils.spec.ts [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+            help: "Critical threshold for suite 'Suite_Context_Mock'"
+        });
+        expect(setMock).toHaveBeenCalledWith(99);
+    });
+
+    it("should register case critical threshold gauge", () =>{
+
+        //GIVEN
+        const contextMock = mockPartial<TestCaseContext>({
+            id: "Case_Context_Mock",
```
Comment (s1hofmann) [inline packages/sakuli-forwarder-prometheus/src/gauge.utils.spec.ts:164]: Are the underscore here wanted?

**Review by s1hofmann — CHANGES_REQUESTED** [review]
A nice and clean implementation.
I added little remarks, but overall it looks solid 👍

File: packages/sakuli-forwarder-prometheus/src/gauge.utils.ts (hunk 2 above)
Comment (svettwer) [reply packages/sakuli-forwarder-prometheus/src/gauge.utils.ts:201]: I haven't recognized any complains from typescript. Strange... :thinking: 
Anyways, I'v double checked with Sakuli core. Sakuli core also uses ES2017 currently.

File: packages/sakuli-forwarder-prometheus/src/prometheus-forwarder.class.ts (hunk 1 above)
Comment (svettwer) [reply packages/sakuli-forwarder-prometheus/src/prometheus-forwarder.class.ts]: Done! Nice catch! :tada:

File: packages/sakuli-forwarder-prometheus/src/prometheus-forwarder.class.ts (hunk 3 above)
Comment (svettwer) [reply packages/sakuli-forwarder-prometheus/src/prometheus-forwarder.class.ts]: Thx! :+1:

File: packages/sakuli-forwarder-prometheus/src/gauge.utils.spec.ts (hunk 4 above)
Comment (svettwer) [reply packages/sakuli-forwarder-prometheus/src/gauge.utils.spec.ts:164]: Yep, wanted to add a little bit variation to test data there. ;-)

**Review by s1hofmann — APPROVED** [review]

**Pull request merged**